000-mm2-connector-offsets-support.md

```diff
@@ -0,0 +1,449 @@
+      "offset": {
+        "offset": 0
+      }
+    }
+  ]
+}
+```
+
+### New annotations
+
+All three actions will make use of a new annotation called `strimzi.io/connector-offsets`.
+The possible values will be `list`, `alter`, and `reset`.
+
+A second new annotation will be added called `strimzi.io/mirrormaker-connector`.
+It will be required when `strimzi.io/connector-offsets` is set.
+The value will be the name of the connector to apply the action to, for example `east-kafka->west-kafka.MirrorSourceConnector`.
+
+If the `strimzi.io/mirrormaker-connector` annotation is set, but the `strimzi.io/connector-offsets` annotation is missing the operator will take no action but add a warning message.
+
+### Listing offsets
+
+To list the current offsets the user will annotate the KafkaMirrorMaker2 resource with the new annotation `strimzi.io/connector-offsets` set to `list`, and the `strimzi.io/mirrormaker-connector` annotation.
+After the annotations are added, on the next reconciliation the operator will fetch the current offsets for the connector and create a Kubernetes ConfigMap containing the JSON response.
+The operator will add an owner reference to the Kubernetes ConfigMap pointing to the KafkaMirrorMaker2 resource that was annotated.
+If the ConfigMap already exists the operator will patch the ConfigMap with updated data and not add an owner reference.
+This means any existing keys in the ConfigMap that the operator is not updating will remain as before.
+Once all list operation is complete the operator will then remove both the `strimzi.io/mirrormaker-connector` and `strimzi.io/connector-offsets` annotations from the KafkaMirrorMaker2 CR.
+
+If the user wants to see an updated set of offsets they will need to re-annotate the resource.
+
+The name of the ConfigMap the operator creates/updates will be set by the user in the KafkaMirrorMaker2 CR.
+The KafkaMirrorMaker2 CRD will be updated to contain a new connector property called `listOffsets` that must be set for the `strimzi.io/connector-offsets=list` annotation to take effect.
+The structure of the `listOffsets` property will be as below, to allow it to be extended in future if required:
+
+```yaml
+spec:
+  #...
+  mirrors:
+    - sourceConnector:
+        #...
+        listOffsets:
+          toConfigMap:
+            name: my-connector-offsets
+      checkpointConnector:
+        #...
+        listOffsets:
+          toConfigMap:
+            name: my-connector-offsets
+      heartbeatConnector:
+        #...
+        listOffsets:
+          toConfigMap:
+            name: my-connector-offsets
+```
+
+The ConfigMap will contain a single entry, containing the response received from the GET /connectors/{connector}/offsets endpoint.
+The name of the entry will be based on the connector name, however since 
+[`>` characters are not allowed in ConfigMap keys](https://kubernetes.io/docs/concepts/configuration/configmap/#configmap-object) the name will be altered to replace `->` with `--`.
```

Comment:
Connector names can include other special characters that are not allows in ConfigMap keys. Will they also get replaced? What happens if there's a collision?

Reply:
This is really unfortunate because I believe Strimzi puts the `->` characters in the connector names.

Reply:
Yeah it is Strimzi that puts the `->` character in. As far as I understand it Strimzi doesn't allow users to specify the connector names for MirrorMaker. It puts the name as `SOURCE_ALIAS->TARGET_ALIAS.CONNECTOR` and the alias field has the regex `"^[a-zA-Z0-9\\._\\-]{1,100}$"` so I don't think we will have a problem. @scholzj do you agree?

Reply:
Ah ok, then the only problematic character should be `>`. I did not know Strimzi had restrictions on the cluster aliases. Thanks